### PR TITLE
drivers: Implement delete methods for S3 and GS

### PR DESCRIPTION
**Pull request by victorges**
Right now it's not in the public Session interface and clients have to hackily cast the sessions. Sue me.

### CODE REVIEW

File: drivers/s3.go [hunk 1]
```diff
@@ -372,6 +373,22 @@ func (os *s3Session) saveDataPut(ctx context.Context, name string, data io.Reade
 	return url, nil
 }
 
+func (os *s3Session) Delete(ctx context.Context, name string) error {
```
Comment (victorges) [inline drivers/s3.go:376]: I see this method is not in the main object store interface. I converted it manually on the hack script where I used it. We can either add it to the main interface and return "not implemented" for other drivers, or even make the "object store deleter" public but still optional and have clients convert stuff to check capabilities. wdyt?

File: drivers/fs.go [hunk 2]
```diff
@@ -121,6 +121,10 @@ func (ostore *FSSession) ListFiles(ctx context.Context, dir, delim string) (Page
 	return pi, nil
 }
 
+func (ostore *FSSession) DeleteFile(ctx context.Context, name string) error {
+	return os.Remove(name)
```
Comment (victorges) [inline drivers/fs.go:125]: Need to join name with prefix here as well

File: drivers/gs.go [hunk 3]
```diff
@@ -156,6 +156,20 @@ func (os *gsSession) createClient() error {
...
+	if !os.useFullAPI {
+		return errors.New("delete not supported for non full api")
+	}
+	if os.client == nil {
+		if err := os.createClient(); err != nil {
+			return err
+		}
+	}
+	return os.client.Bucket(os.bucket).
+		Object(name).
```
Comment (victorges) [inline drivers/gs.go:169]: We should probably join the session key with the filename provided here as well for consistency with other methods

File: drivers/s3.go [hunk 4]
```diff
@@ -372,6 +373,22 @@ func (os *s3Session) saveDataPut(ctx context.Context, name string, data io.Reade
 	return url, nil
 }
 
+func (os *s3Session) DeleteFile(ctx context.Context, name string) error {
+	if os.s3svc == nil {
+		return errors.New("delete not supported for non full api")
+	}
+	params := &s3.DeleteObjectInput{
+		Bucket: aws.String(os.bucket),
+		Key:    aws.String(name),
```
Comment (victorges) [inline drivers/s3.go:382]: Same for joining session key

File: drivers/fs.go (hunk 2 above)
Comment (thomshutt) [reply drivers/fs.go]: Thanks, I'd forgotten that it was based on the base. I've updated the unit test too.

File: drivers/s3.go (hunk 4 above)
Comment (thomshutt) [reply drivers/s3.go:382]: Doesn't this already happen on the lines below?

File: drivers/gs.go (hunk 3 above)
Comment (thomshutt) [reply drivers/gs.go]: Done (I think). Let me know if that's correct

File: drivers/s3.go (hunk 4 above)
Comment (victorges) [reply drivers/s3.go:382]: It has a little different logic, actually only using the key if the provided name is empty. I think we should copy this logic instead: https://github.com/livepeer/go-tools/blob/923fecd7eec57802b8037ee014026430e6b2e03a/drivers/s3.go#L303-L306

File: drivers/s3.go [hunk 5]
```diff
@@ -372,6 +373,22 @@ func (os *s3Session) saveDataPut(ctx context.Context, name string, data io.Reade
 	return url, nil
 }
 
+func (os *s3Session) DeleteFile(ctx context.Context, name string) error {
```
Comment (leszko) [inline drivers/s3.go:376]: You can consider adding tests unit tests for this part, like this one: https://github.com/livepeer/go-tools/blob/44cbd85855e79c9a8d9327598e0a95d9261c0351/drivers/s3_test.go#L60

I know they are really unit tests, but being pragmatic, they are super helpful to check if everything works fine.

**Review by leszko — APPROVED** [review]
Added some comments. Mostly minor. Other than that, LGTM.

File: drivers/ipfs.go [hunk 6]
```diff
@@ -107,6 +109,10 @@ func (session *IpfsSession) GetInfo() *OSInfo {
 	return nil
 }
 
+func (ostore *IpfsSession) DeleteFile(ctx context.Context, name string) error {
+	return errors.New("unsupported method: cannot delete files from IPFS")
```
[the remark this answers is not in the document]
Comment (thomshutt) [reply drivers/ipfs.go:113]: Done

File: drivers/local.go [hunk 7]
```diff
@@ -81,6 +81,10 @@ func (ostore *MemorySession) EndSession() {
 	ostore.os.lock.Unlock()
 }
 
+func (ostore *MemorySession) DeleteFile(ctx context.Context, name string) error {
+	return errors.New("unsupported method: cannot delete files from in-memory session")
```
[the remark this answers is not in the document]
Comment (thomshutt) [reply drivers/local.go:85]: Done

File: drivers/s3.go (hunk 4 above)
Comment (thomshutt) [reply drivers/s3.go:382]: Gotcha, have copied that logic now

File: drivers/s3.go (hunk 5 above)
Comment (thomshutt) [reply drivers/s3.go:376]: Good call, added one and confirmed it works

**Pull request merged**